### PR TITLE
Add unit tests for taskrun

### PR DESCRIPTION
**Pull request by shashwathi**
- test to verify build status update
- test to error when build fetch failure

Address issue: https://github.com/knative/build-pipeline/issues/140

FYI:  I could not prepend reactors to build lister so I have used build client in taskrun to fetch build information. If somebody has another idea feel free to pitch in.

### CODE REVIEW

File: pkg/reconciler/v1alpha1/taskrun/taskrun.go [hunk 1]
```diff
@@ -156,7 +156,7 @@ func (c *Reconciler) Reconcile(ctx context.Context, key string) error {
 
 func (c *Reconciler) reconcile(ctx context.Context, tr *v1alpha1.TaskRun) error {
 	// get build the same as the taskrun, this is the value we use for 1:1 mapping and retrieval
-	build, err := c.buildLister.Builds(tr.Namespace).Get(tr.Name)
+	build, err := c.BuildClientSet.BuildV1alpha1().Builds(tr.Namespace).Get(tr.Name, metav1.GetOptions{})
 	if errors.IsNotFound(err) {
```
Comment (pivotal-nader-ziada) [inline pkg/reconciler/v1alpha1/taskrun/taskrun.go]: if the buildLister is not needed anymore, can we remove it?

File: pkg/reconciler/v1alpha1/taskrun/taskrun_test.go [hunk 2]
```diff
@@ -149,10 +155,19 @@ func TestReconcileBuildsCreated(t *testing.T) {
...
 			}
-			build := client.Actions()[0].(ktesting.CreateAction).GetObject().(*buildv1alpha1.Build)
+			namespace, name, err := cache.SplitMetaNamespaceKey(tc.taskRun)
+			if err != nil {
+				t.Errorf("Invalid resource key: %v", err)
+			}
+			// check error
+			build, err := client.bclient.BuildV1alpha1().Builds(namespace).Get(name, metav1.GetOptions{})
+			//build := client.Actions()[0].(ktesting.CreateAction).GetObject().(*buildv1alpha1.Build)
+			if err != nil {
```
Comment (pivotal-nader-ziada) [inline pkg/reconciler/v1alpha1/taskrun/taskrun_test.go]: remove comment?

File: pkg/reconciler/v1alpha1/taskrun/taskrun_test.go [hunk 3]
```diff
@@ -197,18 +211,136 @@ func TestReconcileBuildCreationErrors(t *testing.T) {
...
+	}
+	build.Status.SetCondition(buildSt)
+
+	_, err := clients.bclient.BuildV1alpha1().Builds(taskRun.Namespace).Create(build)
+	if err != nil {
+		t.Errorf("error creating build : %v", err)
+	}
+
+	if err := c.Reconciler.Reconcile(context.Background(), fmt.Sprintf("%s/%s", taskRun.Namespace, taskRun.Name)); err != nil {
+		t.Fatal("Unexpected error", err)
```
Comment (tejal29) [inline pkg/reconciler/v1alpha1/taskrun/taskrun_test.go]: should this be fatalf?

File: pkg/reconciler/v1alpha1/taskrun/taskrun_test.go [hunk 4]
```diff
@@ -197,18 +211,136 @@ func TestReconcileBuildCreationErrors(t *testing.T) {
...
+
+	// update build status and trigger reconcile
+	buildSt.Status = corev1.ConditionTrue
+	buildSt.Message = "Build completed"
+
+	build.Status.SetCondition(buildSt)
+
+	_, err = clients.bclient.BuildV1alpha1().Builds(taskRun.Namespace).Update(build)
+	if err != nil {
+		t.Errorf("Unexpected error creating build: %v", err)
```
Comment (tejal29) [inline pkg/reconciler/v1alpha1/taskrun/taskrun_test.go]: t.Errorf("unexpected error while updating build: %v", err)

File: pkg/reconciler/v1alpha1/taskrun/taskrun_test.go [hunk 5]
```diff
@@ -197,18 +211,136 @@ func TestReconcileBuildCreationErrors(t *testing.T) {
...
+
+	if err := c.Reconciler.Reconcile(context.Background(), fmt.Sprintf("%s/%s", taskRun.Namespace, taskRun.Name)); err != nil {
+		t.Fatalf("Unexpected error when Reconcile(): %v", err)
+	}
+
+	newTr, err = clients.taskrunInformer.Lister().TaskRuns(taskRun.Namespace).Get(taskRun.Name)
+	if err != nil {
+		t.Errorf("Unexpected error fetching taskrun: %v", err)
+	}
+	if d := cmp.Diff(newTr.Status.GetCondition(duckv1alpha1.ConditionSucceeded), buildSt, ignoreLastTransitionTime); d != "" {
```
Comment (tejal29) [inline pkg/reconciler/v1alpha1/taskrun/taskrun_test.go]: this should not be executed if error is not nil. Maybe use t.Fatalf before or do this in else.

File: pkg/reconciler/v1alpha1/taskrun/taskrun_test.go [hunk 6]
```diff
@@ -197,18 +211,136 @@ func TestReconcileBuildCreationErrors(t *testing.T) {
...
+
+	build.Status.SetCondition(buildSt)
+
+	_, err = clients.bclient.BuildV1alpha1().Builds(taskRun.Namespace).Update(build)
+	if err != nil {
+		t.Errorf("Unexpected error creating build: %v", err)
+	}
+
+	if err := c.Reconciler.Reconcile(context.Background(), fmt.Sprintf("%s/%s", taskRun.Namespace, taskRun.Name)); err != nil {
+		t.Fatalf("Unexpected error when Reconcile(): %v", err)
```
Comment (tejal29) [inline pkg/reconciler/v1alpha1/taskrun/taskrun_test.go]: can we change it to smaller case "unexpected" here and everywhere else

Comment (bobcatfish) [reply pkg/reconciler/v1alpha1/taskrun/taskrun_test.go]: @tejal29 I think that's the standard for errors, but maybe not for test failures?

**Comment by tejal29** [comment]
> @tejal29 @shashwathi The lister returns objects in the cache, while the client makes an actual call to the api to get an object. Things gets added to the cache on a periodic cycle.

ah, then it makes sense to use the client instead of Lister especially since if build was created in last re-concile loop 30 secs ago and it wasn't added to the cache?
WDYT?

File: pkg/reconciler/v1alpha1/taskrun/taskrun_test.go (hunk 3 above)
Comment (shashwathi) [reply pkg/reconciler/v1alpha1/taskrun/taskrun_test.go]: Yes. If reconcile() has failures then no point continuing the test. I will update the error msg

**Comment by shashwathi** [comment]
@tejal29 : I have addressed your comments.

File: pkg/reconciler/v1alpha1/taskrun/taskrun.go (hunk 1 above)
Comment (bobcatfish) [reply pkg/reconciler/v1alpha1/taskrun/taskrun.go]: Did we decide if we want to use the client set here? I think we want to keep using the lister b/c it has cacheing?

File: pkg/reconciler/v1alpha1/taskrun/taskrun_test.go (hunk 2 above)
Comment (bobcatfish) [reply pkg/reconciler/v1alpha1/taskrun/taskrun_test.go]: bump ^^

File: pkg/reconciler/v1alpha1/taskrun/taskrun.go (hunk 1 above)
Comment (bobcatfish) [reply pkg/reconciler/v1alpha1/taskrun/taskrun.go]: Oh i see, there is a discussion below

**Comment by bobcatfish** [comment]
> ah, then it makes sense to use the client instead of Lister especially since if build was created in last re-concile loop 30 secs ago and it wasn't added to the cache?

* Do we know how frequently items are added to the cache?
* If this is the case, do we _ever_ want to use the Lister?
* Lastly, since we are using the Tracker, won't the reconcile get called more frequently than 30 sec. anyway?

I guess it comes down to what mechanism refreshes the objects in the cache. I would think that if this controller _made_ the object (i.e. the build), it would get added to its cache?)

Adding the missing:
/approve

Just in case there's something more about this lister cache business, plz remove if I'm totally offbase:
/hold

**Comment by pivotal-nader-ziada** [comment]
Since the build is created by the controller, we are not adding it to the informer cache, so its probably just safer to use the client

**Comment by bobcatfish** [comment]
> Since the build is created by the controller, we are not adding it to the informer cache, so its probably just safer to use the client

I'm trying to understand how this would work, and looking at [the SharedInformer interface](https://github.com/knative/build-pipeline/blob/master/vendor/k8s.io/client-go/tools/cache/shared_informer.go#L34), it says:

> When you receive a notification, the cache will be AT LEAST as fresh as the notification, but it MAY be more fresh.  

Since we tell the Tracker that we want the reconcile loop to be called when Builds we have created are updated, I would think that this indicates that if the event handler is being triggered, the informer cache will contain the state of the Build as of that update?

**Comment by pivotal-nader-ziada** [comment]
> Since we tell the Tracker that we want the reconcile loop to be called when Builds we have created are updated, I would think that this indicates that if the event handler is being triggered, the informer cache will contain the state of the Build as of that update?

yes, I believe you are correct and the build will get added to the informer, both ways should work, I meant to say the client is guaranteed to find the build before it even gets added to the cache

File: pkg/reconciler/v1alpha1/taskrun/taskrun_test.go (hunk 2 above)
Comment (shashwathi) [reply pkg/reconciler/v1alpha1/taskrun/taskrun_test.go]: Comment is deleted.

**Comment by shashwathi** [comment]
I agree with @pivotal-nader-ziada that client will fetch the recent copy of build. I will remove `hold`

/hold cancel

**Pull request merged**